File: c++/0445.cpp
// Link: https://leetcode.com/problems/add-two-numbers-ii/description/
// Runtime: 46 ms
// Memory: 73.6 MB

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution 
{
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) 
    {
        ListNode* list1 = l1, *list2 = l2;
        std::stack<short> valuesList1, valuesList2;
        while (list1 != nullptr)
        {
            valuesList1.push(list1->val);
            list1 = list1->next;
        } 

        while (list2 != nullptr)
        {
            valuesList2.push(list2->val);
            list2 = list2->next;
        }

        ListNode* start = nullptr;
        short totalSum = 0;
        while ((!valuesList1.empty()) || (!valuesList2.empty()))
        {
            if (!valuesList1.empty()) 
            {
                totalSum += valuesList1.top();
                valuesList1.pop();
            }

            if (!valuesList2.empty()) 
            {
                totalSum += valuesList2.top();
                valuesList2.pop();
            }

            ListNode* newNode = new ListNode(totalSum % 10, start);
            start = newNode;
            totalSum /= 10;

            if ((totalSum != 0) && (valuesList1.empty()) && (valuesList2.empty()))
            {
                ListNode* remainderNode = new ListNode(totalSum, start);
                start = remainderNode;
            }
        }

        return start;
    }
};

File: python/0445.py
# Link: https://leetcode.com/problems/add-two-numbers-ii/description/
# Runtime: 87 ms
# Memory: 16.5 MB

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        list1, list2 = l1, l2
        stack_for_list1, stack_for_list2 = [], []
        while list1 is not None:
            stack_for_list1.append(list1.val)
            list1 = list1.next

        while list2 is not None:
            stack_for_list2.append(list2.val)
            list2 = list2.next

        start = None
        total_sum = 0
        while (len(stack_for_list1) > 0) or (len(stack_for_list2) > 0):
            if len(stack_for_list1) > 0:
                total_sum += stack_for_list1.pop()

            if len(stack_for_list2) > 0:
                total_sum += stack_for_list2.pop()

            new_node = ListNode(total_sum % 10, start)
            total_sum //= 10
            start = new_node

            if (
                (len(stack_for_list1) == 0)
                and (len(stack_for_list2) == 0)
                and (total_sum != 0)
            ):
                last_node = ListNode(total_sum, start)
                start = last_node

        return start
